packages/swap/src/quoting/Quoter.ts

```diff
@@ -0,0 +1,243 @@
+      ...leg.toPoolJSON(),
+      amount: remainingAmount,
+    });
+
+    return swappedAmount + outputAmount;
+  }
+
+  return swappedAmount;
+};
+
+export const approximateIntermediateOutput = async (asset: InternalAsset, amount: string) => {
+  const price = await getAssetPrice(asset);
+
+  if (typeof price !== 'number') return null;
+
+  return BigInt(
+    new BigNumber(amount)
+      .times(price)
+      .shiftedBy(assetConstants.Usdc.decimals - assetConstants[asset].decimals)
+      .toFixed(0),
+  );
+};
+
+export const differenceExceedsThreshold = (
+  a: bigint,
+  b: bigint,
+  tolerance = env.QUOTE_APPROXIMATION_THRESHOLD,
+) => percentDifference(a.toString(), b.toString()).gt(tolerance);
+
+export default class Quoter {
+  private readonly quotes$ = new Subject<Quote>();
+
+  constructor(
+    private readonly io: Server,
+    private createId: () => string = randomUUID,
+  ) {
+    io.on('connection', (socket: QuotingSocket) => {
+      logger.info(`market maker "${socket.data.marketMaker}" connected`);
+
+      socket.on('disconnect', () => {
+        logger.info(`market maker "${socket.data.marketMaker}" disconnected`);
+      });
+
+      socket.on('quote_response', (message) => {
+        const result = marketMakerResponseSchema.safeParse(message);
+
+        if (!result.success) {
+          logger.warn(
+            `received invalid quote response from "${socket.data.marketMaker}"`,
+            {},
+            { message },
+          );
+          return;
+        }
+
+        this.quotes$.next({ marketMaker: socket.data.marketMaker, quote: result.data });
+      });
+    });
+  }
+
+  canQuote() {
+    return env.USE_JIT_QUOTING && this.io.sockets.sockets.size > 0;
+  }
+
+  private async collectMakerQuotes(requestId: string): Promise<MarketMakerQuote[]> {
+    const connectedClients = this.io.sockets.sockets.size;
+    if (connectedClients === 0) return Promise.resolve([]);
+
+    const clientsReceivedQuotes = new Map<string, MarketMakerQuote>();
+
+    return new Promise((resolve) => {
+      let sub: Subscription;
+
+      let timer: ReturnType<typeof setTimeout>;
+
+      const complete = () => {
+        resolve([...clientsReceivedQuotes.values()]);
+        sub.unsubscribe();
+        clearTimeout(timer);
+      };
+
+      sub = this.quotes$
+        .pipe(filter(({ quote }) => quote.request_id === requestId))
+        .subscribe(({ marketMaker, quote }) => {
+          clientsReceivedQuotes.set(marketMaker, quote);
+          if (clientsReceivedQuotes.size === connectedClients) complete();
+        });
+
+      timer = setTimeout(complete, env.QUOTE_TIMEOUT);
+    });
+  }
+
+  private async quoteLegs(legs: [Leg]): Promise<[bigint]>;
+  private async quoteLegs(legs: [Leg, Leg]): Promise<[bigint, bigint]>;
+  private async quoteLegs(legs: [Leg, Leg | null]): Promise<[bigint, bigint | null]>;
+  private async quoteLegs([leg1, leg2]: [Leg] | [Leg, Leg | null]): Promise<
+    [bigint] | [bigint, bigint | null]
+  > {
```

Comment:
does this also account for situations where the mm does not quote the first leg, but quotes the second leg?